app/cmake/ZephyrBuildConfig.cmake

```diff
@@ -74,34 +97,58 @@ foreach(root ${BOARD_ROOT})
 	endif()
 
 	if(DEFINED SHIELD)
-		find_path(shields_refs_list
-		    NAMES ${SHIELD}.overlay
-		    PATHS ${root}/boards/shields/*
-		    NO_DEFAULT_PATH
-		    )
-		foreach(shield_path ${shields_refs_list})
-			get_filename_component(SHIELD_DIR ${shield_path} NAME)
-			list(APPEND KEYMAP_DIRS ${shield_path})
+		foreach(s ${SHIELD_AS_LIST})
+			if(NOT ${s} IN_LIST SHIELD_LIST)
+				message(WARNING "Didn't find ${s}")
+				continue()
+			endif()
+			message(STATUS "Adding ${SHIELD_DIR_${s}}")
+			list(APPEND KEYMAP_DIRS ${SHIELD_DIR_${s}})
+			get_filename_component(shield_dir_name ${SHIELD_DIR_${s}} NAME)
+			list(APPEND SHIELD_DIR ${shield_dir_name})
 		endforeach()
 	endif()
 endforeach()
 
+# Give a shield like `kyria_rev2_left` we want to use `kyria_rev2` and `kyria` as candidate names for
+# overlay/conf/keymap files.
+if(DEFINED SHIELD)
+	foreach(s ${SHIELD_AS_LIST})
+		message(STATUS ${SHIELD_DIR_${s}})
+		get_filename_component(shield_dir_name ${SHIELD_DIR_${s}} NAME)
+		string(REPLACE "_" ";" S_PIECES ${s})
+		list(LENGTH S_PIECES S_PIECES_LEN)
+		foreach (n RANGE ${S_PIECES_LEN-2})
+			list(POP_BACK S_PIECES)
+			list(JOIN S_PIECES "_" S_SUBSTR)
+			if (S_SUBSTR STREQUAL ${shield_dir_name})
+				break()
+			endif()
+			list(APPEND shield_candidate_names ${S_SUBSTR})
+		endforeach()
+	endforeach()
+endif()
+
 if (ZMK_CONFIG)
 	if (EXISTS ${ZMK_CONFIG})
 		message(STATUS "ZMK Config directory: ${ZMK_CONFIG}")
 		list(APPEND DTS_ROOT ${ZMK_CONFIG})
 		list(PREPEND KEYMAP_DIRS "${ZMK_CONFIG}")
 
-		if (SHIELD)
-			message(STATUS "Board: ${BOARD}, ${BOARD_DIR}, ${SHIELD}, ${SHIELD_DIR}")
-			list(APPEND overlay_candidates "${ZMK_CONFIG}/${SHIELD_DIR}.overlay")
-			list(APPEND overlay_candidates "${ZMK_CONFIG}/${SHIELD_DIR}_${BOARD}.overlay")
-			list(APPEND overlay_candidates "${ZMK_CONFIG}/${SHIELD}_${BOARD}.overlay")
-			list(APPEND overlay_candidates "${ZMK_CONFIG}/${SHIELD}.overlay")
-			list(APPEND config_candidates "${ZMK_CONFIG}/${SHIELD_DIR}.conf")
-			list(APPEND config_candidates "${ZMK_CONFIG}/${SHIELD_DIR}_${BOARD}.conf")
-			list(APPEND config_candidates "${ZMK_CONFIG}/${SHIELD}_${BOARD}.conf")
-			list(APPEND config_candidates "${ZMK_CONFIG}/${SHIELD}.conf")
+		if (DEFINED SHIELD)
+			foreach (s ${shield_candidate_names})
+				if (DEFINED $SHIELD_DIR_${s})
+					get_filename_component(shield_dir_name ${SHIELD_DIR_${s}} NAME)
+				endif()
+				list(APPEND overlay_candidates "${ZMK_CONFIG}/${s}_${BOARD}.overlay")
+				list(APPEND overlay_candidates "${ZMK_CONFIG}/${s}.overlay")
+				if (NOT "${shield_dir_name}" STREQUAL "${s}")
+					list(APPEND config_candidates "${ZMK_CONFIG}/${shield_dir_name}.conf")
+					list(APPEND config_candidates "${ZMK_CONFIG}/${shield_dir_name}_${BOARD}.conf")
```

Comment:
Are the candidates searched in order? If so, is there any reason the board-specific version is second here but first in the surrounding cases?

Reply:
No, good catch. Fixed.